src/OrchardCore/OrchardCore.Localization.Abstractions/CultureDictionaryRecordKey.cs

```diff
@@ -0,0 +1,68 @@
+using System;
+
+namespace OrchardCore.Localization
+{
+    /// <summary>
+    /// Represents a key for <see cref="CultureDictionaryRecord"/>.
+    /// </summary>
+    public struct CultureDictionaryRecordKey : IEquatable<CultureDictionaryRecordKey>
+    {
+        private readonly string _messageId;
+        private readonly string _context;
+        private readonly string _key;
+
+        /// <summary>
+        /// Creates new instance of <see cref="CultureDictionaryRecordKey"/>.
+        /// </summary>
+        /// <param name="messageId">The message Id.</param>
+        public CultureDictionaryRecordKey(string messageId) : this(messageId, null)
+        {
+
+        }
+
+        /// <summary>
+        /// Creates new instance of <see cref="CultureDictionaryRecordKey"/>.
+        /// </summary>
+        /// <param name="messageId">The message Id.</param>
+        /// <param name="context">The message context.</param>
+        public CultureDictionaryRecordKey(string messageId, string context)
+        {
+            _messageId = messageId;
+            _context = context;
+
+            _key = String.IsNullOrEmpty(context)
```

Comment:
The goal is to not have to create a new string. You need to remove this and the implicit conversion to string.

Reply:
I was worried about this line before I wrote it, so that's why I asked for your review

Reply:
We need to find a way to provide a key for `IStringLocalizer` implementations

Reply:
Just moving the key string generation when `string` casting is happen (JIT), this may it allocates when we called `GetAllStrings()` but I think it's fine because this call is rare

@sebastienros I just left the implicit string conversion for purpose, I don't think it allocates unless the conversion is occurs

Please if you have a better way shoot it to me ;)